### ISSUE TITLE
Buffers don't seem to appear in heap profiles

### ISSUE THREAD

**Issue by nolanmar511** [GoogleCloudPlatform/cloud-profiler-nodejs#157]
I ran this code:

```
const profiler = require('@google-cloud/profiler')
profiler.start({
	projectId: 'nolanmar-testproject',
	logLevel: 5,
	serviceContext: {
		service: 'buffer-bench',
		version: '0'
	},
});

var bench;

function buffer_bench() {
    bench = new Buffer(1e6);
    bench.fill('');
    setImmediate(buffer_bench);
}

buffer_bench();
```

And the function buffer_bench didn't appear in the heap profile:
![image](https://user-images.githubusercontent.com/14046562/37228357-292ebf64-2395-11e8-946e-4e13f64c93ac.png)

The [docs]( https://nodejs.org/api/buffer.html) for Buffer say "Instances of the Buffer class are similar to arrays of integers but correspond to fixed-sized, raw memory allocations outside the V8 heap. The size of the Buffer is established when it is created and cannot be resized."

Does that mean we would expect to not see the full buffer allocation in the profile?
Can this be fixed?

**Comment by aalexand** [comment]
I don't think these are changing the observations much but some comments on the sample code:
* I'd allocate a power of

**Comment by aalexand** [comment]
@ofrobots PTAL whether this is expected behavior that allocations via Buffer do not seem to be reported by heap profiler.

**Comment by ofrobots** [comment]
Yeah, I guess this makes sense. Buffers are not allocated on the JS heap.

**Comment by ofrobots** [comment]
There may be two ways to fix this:

* Somehow interject into the [ArrayBufferAllocator](https://github.com/nodejs/node/blob/master/src/node.cc#L779) in Node or in V8.
* Introduce our own malloc/free into the process that uses tcmalloc; thereby intercepting *all* native memory.

**Comment by ofrobots** [comment]
There a second problem in the benchmark above. The allocated buffer goes dead very soon after allocation, so GC will collect it fast. At the most, we will have one sample in the heap profile corresponding to the JS wrapper around the native buffer. It is quite likely that we have zero samples.

**Comment by aalexand** [comment]
bench is a global variable so the buffer should stay live? But I agree the sample code is a bit strange:
* 1e6 is not a power of two, is that intentional?
* Which sampling threshold of heap profiler was used in this run?
* Why allocate in a loop? Single large allocation referenced from a global variable should be enough. Shouldn't need the loop at all imo.

**Comment by aalexand** [comment]
Also this issue should have been filed against sampling heap profiler repo imo.

**Comment by ofrobots** [comment]
At any given time there is at most live buffer object anchored in the `bench` object. The question is whether that buffer happened to be lucky enough to get sampled. The javascript object wrapping the native buffer is only ~3 words big, so the probability it would get sampled is ~.000045776.

**Comment by nolanmar511** [comment]
The function which allocates the does not appear in this code:
```
const profiler = require('@google-cloud/profiler')
profiler.start({
	projectId: 'nolanmar-testproject',
	logLevel: 5,
	serviceContext: {
		service: 'buffer-bench',
		version: '2'
	},
});

var buf;

function bufferAlloc() {
	buf = new Buffer(2^20);
}

function bufferFill() { 
   	buf.fill(0)
}

bufferAlloc();
setInterval(bufferFill, 1)
```
Here's the profile:
![image](https://user-images.githubusercontent.com/14046562/37233828-adb17f1a-23a9-11e8-80be-6d9a67a5b8e0.png)

I do

**Comment by ofrobots** [comment]
At the time the profile is taken, there is one currently live `buf` object. The probability that particular buffer object got sampled is ~.000045776.

Try this change instead:
```js
var bufs

function bufferAlloc() {
	bufs.push(new Buffer(2^20));
}
```

**Comment by ofrobots** [comment]
Note that the `2^20` is irrelevant as that corresponds to an allocation on the native, rather than the JS heap. Why my suggested change above, you can probably use a smaller size – else the leak will grow quite fast.

**Comment by aalexand** [comment]
> Note that the 2^20 is irrelevant as that corresponds to an allocation on the native, rather than the JS heap. 

But 2^20 not being sampled / recorded is exactly what this issue is about. I agree that doing a bunch of small buffer allocations will eventually produce some samples because there is some JS heap overhead of each of those allocations, but it's less interesting than fixing the profiler to actually report the buffer allocations.

**Comment by ofrobots** [comment]
Two problems are being reported here:
* "[The Profile] doesn't have bufferAlloc() or bufferFill() appear."
* Backing storage for Buffers cannot be sampled.

I agree the former should be pulled off into a separate issue.

The latter is quite hard, and it is might actually be broader issue that this project presently does not aim to be able to profile native allocations.

**Comment by aalexand** [comment]
Does the "externally allocated memory" metric goes up when a buffer is allocated? If yes and we expose that as discussed in https://github.com/GoogleCloudPlatform/cloud-profiler-nodejs/issues/121 it might be better than nothing.

**Comment by ofrobots** [comment]
Yes, the `external` metric reported by `process.memoryUsage` will track the memory used by native buffers.

**Comment by aalexand** [comment]
@nolanmar511 Please confirm that the buffer allocations are captured in the external metric and once that's hopefully confirmed we should address it first by doing #121. Then once we see that the external memory usage is a significant bucket for any customer or app we use, we can start breaking this down (but this will be a one off thing for every type of native memory allocation perhaps).

**Comment by aalexand** [comment]
> @nolanmar511 Please confirm that the buffer allocations are captured in the external metric 

@nolanmar511 Did you look at this?

**Comment by nolanmar511** [comment]
The external memory usage does appear to contain buffers.

I ran the following, varying the size of the buffer. External memory was always slightly larger (a few KiB for 5MiB buffer) than the buffer.

```
const buf = Buffer.alloc(5 * 1024 * 1024);
console.log(process.memoryUsage());
```

**Comment by aalexand** [comment]
@ofrobots @nolanmar511 Questions:
* Should we name the pseudo node in profiles as "(external)" or as "(buffers)" or something else?
* Should this pseudo node be created in the sampling heap profiler code in Node.js or in cloud-profiler-nodejs code?

**Comment by aalexand** [comment]
@ofrobots @nolanmar511 Thoughts?

**Comment by nolanmar511** [comment]
I have a preference for "(external)", since I believe the external memory can contain things other than buffers.

**Comment by ofrobots** [comment]
+1 on "(external)"

**Comment by Venryx** [comment]
`process.memoryUsage()` also returns an `arrayBuffers` property, if you want to distinguish between all external memory and just the array buffers.
